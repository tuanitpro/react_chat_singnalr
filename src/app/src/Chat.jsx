import React, { Component } from 'react'
import moment from 'moment';
const signalR = require("@aspnet/signalr");

class Chat extends Component {
    constructor(props) {
        super(props);
        this.state = {
            nick: 'Guest',
            isLogin: false,
            message: '',
            messages: [],
            hubConnection: null
        }

        this.login = this.login.bind(this);
        this.sendMessage = this.sendMessage.bind(this);
    }
    componentDidMount() {
        let hubConnection = new signalR.HubConnectionBuilder()
            .configureLogging(signalR.LogLevel.Debug)
            .withUrl("http://localhost:5000/chat", {
                skipNegotiation: false,
                transport: signalR.HttpTransportType.WebSockets | signalR.HttpTransportType.LongPolling
            })
            .build();

        hubConnection.start()
            .then(() => console.log('Connection started!'))
            .catch(err => console.error(err));

        hubConnection.on('sendToAll', (nick, receivedMessage) => {
            let dateTimeNow = new Date();
            let timeAgo = moment(dateTimeNow).format('LTS'); 
            const text = `${nick}: ${receivedMessage} at ${timeAgo}`;
            const messages = this.state.messages.concat([text]);
            this.setState({ messages });
        });

        this.setState({
            hubConnection: hubConnection
        })
    }

    sendMessage() {
        const { hubConnection, nick, message } = this.state;
        hubConnection
            .invoke('sendToAll', nick, message)
            .catch(err => console.error(err));

        this.setState({ message: '' });
    }

    login() {
        const { hubConnection, nick } = this.state;
        this.setState({ isLogin: true });
        let message = `has joined`;
        hubConnection
            .invoke('sendToAll', nick, message)
            .catch(err => console.error(err));
        this.setState({ message: '' });
    }

    render() {
        const { isLogin, nick, message, messages } = this.state;
        return (
            <div>
                {!isLogin &&
                    <div className="login-form">
                        <input
                            type="text"
                            placeholder="Nick"
                            value={nick}
                            onChange={e => this.setState({ nick: e.target.value })}
                        />
                        <button onClick={this.login}>Login</button>
                    </div>
                }
                {isLogin &&
                    <div className="message-board">
                        <input
                            type="text"
                            placeholder="Message"
                            value={message}
                            onChange={e => this.setState({ message: e.target.value })}
                        />

                        <button onClick={this.sendMessage}>Send</button>

                        <div>
                            {messages.map((message, index) => (
                                <span style={{ display: 'block' }} key={index}> {message} </span>
                            ))}
                        </div>
                    </div>
                }
            </div>
        )
    }
}

export default Chat