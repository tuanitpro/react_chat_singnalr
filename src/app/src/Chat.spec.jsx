import React from 'react';
import Chat from './Chat';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { shallow } from 'enzyme';
Enzyme.configure({ adapter: new Adapter() });

const signalR = require("@aspnet/signalr");
// jest.mock('signalR');

describe('Chat component', () => {
    it('render', (done) => {
        const component = shallow(<Chat />);
        expect(component).toMatchSnapshot();
    });

})