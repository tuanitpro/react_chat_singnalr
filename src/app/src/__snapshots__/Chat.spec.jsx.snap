// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat component render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Chat />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="login-form"
        >
          <input
            onChange={[Function]}
            placeholder="Nick"
            type="text"
            value="Guest"
          />
          <button
            onClick={[Function]}
          >
            Login
          </button>
        </div>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              onChange={[Function]}
              placeholder="Nick"
              type="text"
              value="Guest"
            />,
            <button
              onClick={[Function]}
            >
              Login
            </button>,
          ],
          "className": "login-form",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onChange": [Function],
              "placeholder": "Nick",
              "type": "text",
              "value": "Guest",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Login",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Login",
            "type": "button",
          },
        ],
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="login-form"
          >
            <input
              onChange={[Function]}
              placeholder="Nick"
              type="text"
              value="Guest"
            />
            <button
              onClick={[Function]}
            >
              Login
            </button>
          </div>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                onChange={[Function]}
                placeholder="Nick"
                type="text"
                value="Guest"
              />,
              <button
                onClick={[Function]}
              >
                Login
              </button>,
            ],
            "className": "login-form",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onChange": [Function],
                "placeholder": "Nick",
                "type": "text",
                "value": "Guest",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Login",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Login",
              "type": "button",
            },
          ],
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
